File: jcv-core/src/main/java/com/ekino/oss/jcv/core/initializer/Initializers.java
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core.initializer;

import java.util.Optional;
import java.util.stream.Stream;

import com.ekino.oss.jcv.core.JsonValidator;
import com.ekino.oss.jcv.core.validator.DefaultParameterizedTemplateValidator;
import com.ekino.oss.jcv.core.validator.ValidatorTemplateManager;

public class Initializers {

    /**
     * Prepare "parameterized" validator for a given comparator.
     *
     * @param id                    the validator id
     * @param comparatorInitializer the comparator init with the parameter
     * @param <T>                   the field value type
     *
     * @return the validator
     */
    public static <T> JsonValidator<T> parameterizedValidator(String id,
                                                              TemplatedComparatorInitializer<T> comparatorInitializer) {
        return new DefaultParameterizedTemplateValidator<>(id, comparatorInitializer);
    }

    public static <T> TemplatedComparatorInitializer<T> comparatorWithoutParameter(NoParameterComparatorInitializer<T> comparatorInitializer) {
        return validatorTemplateManager -> comparatorInitializer.initComparator();
    }

    public static <T> TemplatedComparatorInitializer<T> comparatorWith1Parameter(OneParameterComparatorInitializer<T> comparatorInitializer) {
        return comparatorWith1Parameter(true, comparatorInitializer);
    }

    public static <T> TemplatedComparatorInitializer<T> comparatorWith1Parameter(boolean required, OneParameterComparatorInitializer<T> comparatorInitializer) {
        return validatorTemplateManager -> {
            String parameter = getOrThrowParameter(0, required, validatorTemplateManager);
            return comparatorInitializer.initComparator(parameter);
        };
    }

    public static <T> TemplatedComparatorInitializer<T> comparatorWith2Parameters(boolean param1Required, boolean param2Required, TwoParametersComparatorInitializer<T> comparatorInitializer) {
        return validatorTemplateManager -> {
            String parameter1 = getOrThrowParameter(0, param1Required, validatorTemplateManager);
            String parameter2 = getOrThrowParameter(1, param2Required, validatorTemplateManager);
            return comparatorInitializer.initComparator(parameter1, parameter2);
        };
    }

    private static String getOrThrowParameter(int index, boolean required, ValidatorTemplateManager validatorTemplateManager) {
        Optional<String> parameter = validatorTemplateManager.extractParameter(index);
        if (required && !parameter.isPresent()) {
            throw new IllegalArgumentException(String.format(
                "No parameter at index %s found in validator '%s'",
                index,
                validatorTemplateManager.extractTemplateContent()
            ));
        }

        return parameter.orElse(null);
    }

    @SafeVarargs
    public static <T> TemplatedComparatorInitializer<T> allOf(TemplatedComparatorInitializer<T>... initializers) {
        return validatorTemplateManager -> (actual, expected) -> Stream.of(initializers)
            .map(it -> it.initComparator(validatorTemplateManager))
            .allMatch(it -> it.equal(actual, expected));
    }

    @SafeVarargs
    public static <T> TemplatedComparatorInitializer<T> anyOf(TemplatedComparatorInitializer<T>... initializers) {
        return validatorTemplateManager -> (actual, expected) -> Stream.of(initializers)
            .map(it -> it.initComparator(validatorTemplateManager))
            .anyMatch(it -> it.equal(actual, expected));
    }
}

File: jcv-core/src/main/kotlin/com/ekino/oss/jcv/core/initializer/Initializers.kt
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core.initializer

import com.ekino.oss.jcv.core.JsonValidator
import com.ekino.oss.jcv.core.validator.DefaultParameterizedTemplateValidator
import com.ekino.oss.jcv.core.validator.ValidatorTemplateManager
import org.skyscreamer.jsonassert.ValueMatcher

object Initializers {

  /**
   * Prepare "parameterized" validator for a given comparator.
   *
   * @param id the validator id
   * @param initializer the comparator init with the parameter
   * @param <T>                   the field value type
   *
   * @return the validator
   */
  @JvmStatic
  fun <T> parameterizedValidator(
    id: String,
    initializer: TemplatedComparatorInitializer<T>,
  ): JsonValidator<T> =
    DefaultParameterizedTemplateValidator(id, initializer)

  @JvmStatic
  fun <T> comparatorWithoutParameter(initializer: NoParameterComparatorInitializer<T>): TemplatedComparatorInitializer<T> =
    TemplatedComparatorInitializer<T> { initializer.initComparator() }

  @JvmStatic
  @JvmOverloads
  fun <T> comparatorWith1Parameter(
    required: Boolean = true,
    initializer: OneParameterComparatorInitializer<T>,
  ): TemplatedComparatorInitializer<T> =
    TemplatedComparatorInitializer<T> { validatorTemplateManager ->
      val parameter = getOrThrowParameter(0, required, validatorTemplateManager)
      initializer.initComparator(parameter)
    }

  @JvmStatic
  fun <T> comparatorWith2Parameters(
    param1Required: Boolean = true,
    param2Required: Boolean = true,
    initializer: TwoParametersComparatorInitializer<T>,
  ): TemplatedComparatorInitializer<T> {
    return TemplatedComparatorInitializer<T> { validatorTemplateManager ->
      val parameter1 = getOrThrowParameter(0, param1Required, validatorTemplateManager)
      val parameter2 = getOrThrowParameter(1, param2Required, validatorTemplateManager)
      initializer.initComparator(parameter1, parameter2)
    }
  }

  private fun getOrThrowParameter(
    index: Int,
    required: Boolean,
    validatorTemplateManager: ValidatorTemplateManager,
  ): String? {
    val parameter = validatorTemplateManager.extractParameter(index)
    require(!(required && parameter == null)) {
      "No parameter at index $index found in validator '${validatorTemplateManager.extractTemplateContent()}'"
    }
    return parameter
  }

  @JvmStatic
  @SafeVarargs
  fun <T> allOf(vararg initializers: TemplatedComparatorInitializer<in T>): TemplatedComparatorInitializer<T> =
    TemplatedComparatorInitializer<T> { validatorTemplateManager ->
      ValueMatcher { actual, expected ->
        sequenceOf(*initializers)
          .map { it.initComparator(validatorTemplateManager) }
          .all { it.equal(actual, expected) }
      }
    }

  @JvmStatic
  @SafeVarargs
  fun <T> anyOf(vararg initializers: TemplatedComparatorInitializer<in T>): TemplatedComparatorInitializer<T> =
    TemplatedComparatorInitializer<T> { validatorTemplateManager ->
      ValueMatcher { actual, expected ->
        sequenceOf(*initializers)
          .map { it.initComparator(validatorTemplateManager) }
          .any { it.equal(actual, expected) }
      }
    }
}
